BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h

```diff
@@ -0,0 +1,153 @@
+//
+//  BOXURLSessionCacheClient.h
+//  BoxContentSDK
+//
+//  Created by Thuy Nguyen on 3/3/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@interface BOXURLSessionTaskCachedInfo : NSObject
+
+@property (nonatomic, strong, readwrite) NSString *backgroundSessionId;
+@property (nonatomic, assign, readwrite) NSUInteger sessionTaskId;
+@property (nonatomic, strong, readwrite) NSString *destinationFilePath;
+@property (nonatomic, strong, readwrite) NSData *resumeData;
+@property (nonatomic, strong, readwrite) NSURLResponse *response;
+@property (nonatomic, strong, readwrite) NSData *responseData;
+@property (nonatomic, strong, readwrite) NSError *error;
+
+@end
+
+@protocol BOXURLSessionCacheClientDelegate <NSObject>
+
+// allow delegate to encrypt data before BOXURLSessionCacheClient persists it to disk
+- (NSData *)encryptData:(NSData *)data;
+
+// allow delegate to decrypt data which was encrypted by encryptData method above
+- (NSData *)decryptData:(NSData *)data;
+
+@end
+
+/**
+ * BOXURLSessionCacheClient is used to persist data for background download/upload session tasks to provide to task delegates
+ * for processing once the task delegates are ready.
+ *
+ * BOXURLSessionCacheClient will require a root directory to cache its data at initialization, and it will create the sub directory
+ * structures to persist data accordingly.
+ *
+ * 1. To store data specific to a background session task uniquely identified by background session id and session task id,
+ * we use sub-directory /sessions/$backgroundSessionId/$sessionTaskId, which will contain up to 5 files:
+ *
+ * - destinationFilePath: store the destination file path of downloaded file, applicable for background download task only
+ * - resumeData:          store the resume data of background download task which allows us to resume
+ *                        download from the point before the task was cancelled, applicable for background
+ *                        download task only
+ * - responseData:        store the response data (could contain error from server)
+ * - response:            store NSURLResponse
+ * - error:               store client-side NSError
+ *
+ * 2. To keep track of whose user and its associateId the session task belongs to, we save backgroundSessionId and sessionTaskId
+ * as a file name under sub-directory /users/$userId/$associateId with format $backgroundSessionId-$sessionTaskId
+ */
+@interface BOXURLSessionCacheClient : NSObject
+
+@property (nonatomic, weak, readwrite) id<BOXURLSessionCacheClientDelegate> delegate;
```

Comment:
Should have a comment explaining that it is OK to leave this unset and that it would just mean the networking cache information will not be encrypted.